### DISCUSSION TITLE
Does Deno.writeTextFile acquire a lock?

### DISCUSSION THREAD

**Discussion by alexgleason** [denoland/deno#26692] [Q&A] [answered]
My original code looked like this:

```ts
await Deno.writeTextFile(path, data);
```

But then it occurred to me that it's really important that multiple Deno processes don't try to write the file at the same time. Thus, **I need to acquire a lock.** So I rewrote the code:

```ts
using file = await Deno.open(path, { write: true, create: true });
await file.lock(true);

const buffer = new TextEncoder().encode(data);

const writer = file.writable.getWriter();
await writer.write(buffer);
await writer.close();
```

But then I wondered, is it possible that `Deno.writeTextFile` already does this for me? It's a convenience function, and it would seem intuitive to acquire a lock while writing. But I couldn't find the answer.

**Comment by 0f-0b** [comment, marked as answer]
It does not, which you can confirm by calling `writeTextFile` while the file is exclusively locked.

By the way, you forgot to `truncate` the file after locking it. In case lock contention turns out to be a problem, another way to rewrite a file atomically is to write all data into a temporary file and then replace the original file using a rename.

```ts
const tempFile = await Deno.makeTempFile({ dir: dirname(path) });
try {
  await Deno.writeTextFile(tempFile, data);
  await Deno.rename(tempFile, path);
} catch (e) {
  await Deno.remove(tempFile);
  throw e;
}
```

**Reply by bartlomieju** [reply]
This is the correct answer. I'll be happy to merge PRs that improve the docstrings on relevant types to make it more clear.
